File: 0002-add-two-numbers/0002-add-two-numbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    
    fun sumListNodes(l1: ListNode?, l2: ListNode?, rest: Int): ListNode? {
        val sum = (l1?.`val` ?: 0) + (l2?.`val` ?: 0) + rest
        val curNode = ListNode(sum % 10)

        if (l1?.next != null || l2?.next != null || sum / 10 != 0) {
            curNode.next = sumListNodes(l1?.next, l2?.next, sum / 10)
        }

        return curNode
    }

    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        return sumListNodes(l1, l2, 0)
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2, rest) {
    var sum = 0;
    
    if (!l1 && !l2) {
        if (rest > 0) {
            return new ListNode(1);
        } else return;
    }
    
    sum = ((l1?.val || 0) + (l2?.val || 0) + (rest || 0)) % 10;
    rest = (l1?.val || 0) + (l2?.val || 0) + (rest || 0) > 9 ? 1 : 0;
    
    l1 = l1 ? l1.next : null;
    l2 = l2 ? l2.next : null;

    return new ListNode(sum, addTwoNumbers(l1, l2, rest));
};